forward expect header

i110 opened the pull request:
Now protocol handlers immediately respond with 100 continue response when it sees `expect: 100-continue` header, and proxy handler (and httpclient) never forward it to the origin, and vice versa (100 continue response never be forwarded to the client). This PR adds `forward` mode to change that behavior in opt-in way.

## Table of key behaviors depending on `proxy.expect` value: 
| `proxy.expect`'s value | OFF | ON | FORWARD |
|---|---|---|---|
| Protocol handlers responds with 100 when it sees `expect` req header? | Yes | Yes | No |
| httpclient adds its own `expect` req header? | No | Yes | No |
| httpclient forwards `expect` req header? | No | No | Yes |
| httpclient suspends sending req body until it sees 100? | No | Yes | No |
| Proxy handler forwards 100 response from upstream? | No | No | Yes |

## TODOs: 
- [x] test

i110 commented:
CI failure will be resolved by merging https://github.com/h2o/h2o/pull/3425 in

i110 commented:
@kazuho I think this is ready for review, could you take a look when you have time?

kazuho commented on include/h2o.h:
```diff
@@ -2551,6 +2569,17 @@ inline int h2o_req_can_stream_request(h2o_req_t *req)
+        if (handlers[i]->supports_expect)
+            return 1;
+    }
+    return 0;
```
I think we'd only want to check the first handler, like `h2o_req_can_stream_request`.

Consider the case where the first handler does not support `supports_expect` but the second handler does. The first handler might block until all the request body arrives. To let that happen, the protocol handlers need to send 100 themselves.

i110 replied:
Addressed in https://github.com/h2o/h2o/pull/3419/commits/71e29f72f07d4dfdcff1adbb073e33c3bdf19335

kazuho commented on include/h2o.h:
```diff
@@ -337,6 +341,12 @@ typedef enum h2o_proxy_zerocopy_mode {
+typedef enum h2o_proxy_expect_mode {
+    H2O_PROXY_EXPECT_DISABLED,
+    H2O_PROXY_EXPECT_ENABLED,
+    H2O_PROXY_EXPECT_FORWARD,
```
Can we have doc-comments explaining how these work? I see the explanation on the PR, but people reading code do not necessarily go back to the original PR (and the PR never gets updated as we change the definition).

i110 replied:
Addressed in https://github.com/h2o/h2o/pull/3419/commits/71e29f72f07d4dfdcff1adbb073e33c3bdf19335

kazuho commented on lib/core/proxy.c:
```diff
@@ -818,7 +818,7 @@ static h2o_httpclient_head_cb on_connect(h2o_httpclient_t *client, const char *e
     if (req->overrides != NULL) {
         use_proxy_protocol = req->overrides->use_proxy_protocol;
-        props->use_expect = req->overrides->proxy_use_expect;
+        props->use_expect = req->overrides->proxy_expect_mode == H2O_PROXY_EXPECT_ENABLED;
```
Do we want to rename `use_expect` to `send_own_expect` or something? I ask this because we now have two ways of "using expect."

i110 replied:
https://github.com/h2o/h2o/pull/3419/commits/eeb8dd01ccb81011cc3c8c66153e41fef9206196

kazuho commented on lib/handler/configurator/proxy.c:
```diff
@@ -618,10 +618,23 @@ static int on_config_http3_ratio(h2o_configurator_command_t *cmd, h2o_configurat
+        self->vars->conf.expect_mode = H2O_PROXY_EXPECT_FORWARD;
+        break;
+    default:
+        return -1;
```
`h2o_fatal` would be a better choice here. We do not want to see the program running in an unexpected state after a logic flaw.

i110 replied:
hmm, but it seems not aligned with the other parts of this file, e.g. `on_config_zerocopy`. Do we have to tweak it as well then?

kazuho replied:
In the two functions that you point out, the path that returns -1 is taken when `h2o_configurator_get_one_of` returns -1 (i.e., when the value in the configuration file does not match any of the provided values).

In the case of the code added in this PR, that check is done before the `switch` statement in question. Therefore this `default` label and the return is excessive.

i110 replied:
ah! I see the point. thank you

kazuho commented on include/h2o.h:
```diff
@@ -134,6 +134,10 @@ typedef struct st_h2o_handler_t {
+    /**
+     *
+     */
+    unsigned supports_expect : 1;
```
```suggestion
    unsigned handles_expect : 1;
```
Maybe a matter of tastes, but I think "handles" makes it clear how the handler "supports" expect.

i110 replied:
Renamed in https://github.com/h2o/h2o/pull/3419/commits/5fcfd4dcb9694c3584e1bcb0cdfae52f7c3cacab

kazuho commented on lib/http1.c:
```diff
@@ -451,8 +453,8 @@ static const char *fixup_request(struct st_h2o_http1_conn_t *conn, struct phr_he
-    if ((ret = init_headers(&conn->req.pool, &conn->req.headers, headers, num_headers, &connection, &host, &upgrade, expect,
-                            entity_header_index)) != NULL)
+    if ((ret = init_headers(&conn->req.pool, &conn->req.headers, headers, num_headers, &connection, &host, &upgrade,
+                            expect, expect_orig_case, entity_header_index)) != NULL)
```
Rather than passing the name and value separately, how about passing the index of the expect header within `headers`?

Then, we can get the name by using `headers[expect_index].name(_len)` and the value by `headers[expect_index].value(_len)`.

i110 replied:
Addressed in https://github.com/h2o/h2o/pull/3419/commits/01fc92a93cf4d8b6f6e34aa65d6a0c10e929a27a

kazuho reviewed (commented):
Thank you for the PR, sorry for the delay. Looks good overall.

kazuho commented on include/h2o.h:
```diff
@@ -1563,6 +1584,10 @@ void h2o_process_request(h2o_req_t *req);
  * returns the first handler that will be used for the request
  */
 h2o_handler_t *h2o_get_first_handler(h2o_req_t *req);
+/**
+ * returns list of handlers and its size that will be used for the request
+ */
+size_t h2o_get_handlers(h2o_req_t *req, h2o_handler_t ***out);
```
```suggestion
```
I think you removed the definition now that it is no longer used, but forgot to remove the declaration here?

kazuho commented on lib/core/request.c:
```diff
@@ -357,8 +357,10 @@ int h2o_req_validate_pseudo_headers(h2o_req_t *req)
+    if (req->pathconf == NULL) {
+        h2o_hostconf_t *hostconf = h2o_req_setup(req);
+        setup_pathconf(req, hostconf);
+    }
```
This looks like a change unrelated to this PR. If it is necessary, could you clarify why we want to make this change?

i110 replied:
It's related. As of this PR, `h2o_get_first_handler` can be called twice (one by h2o_req_can_stream_request, another by h2o_req_should_forward_expect). Unless we add this change, h2o_req_setup and setup_pathconf will be called twice (I don't think it causes any serious failures but it's just wasteful)

kazuho commented on lib/handler/configurator/proxy.c:
```diff
@@ -618,10 +618,21 @@ static int on_config_http3_ratio(h2o_configurator_command_t *cmd, h2o_configurat
+        self->vars->conf.expect_mode = H2O_PROXY_EXPECT_DISABLED;
+        break;
+    case 1:
+        self->vars->conf.expect_mode = H2O_PROXY_EXPECT_ENABLED;
```
Sorry to suggest this late, but would it make sense to call these ~_OFF and ~_ON, considering that they are referred to as such by the config yaml?

i110 replied:
Renamed in https://github.com/h2o/h2o/pull/3419/commits/faf6912fc323f6f546d18b56afba5f247fa7b9f4

kazuho commented on lib/http1.c:
```diff
@@ -532,6 +535,15 @@ static const char *fixup_request(struct st_h2o_http1_conn_t *conn, struct phr_he
 
+    /* in forward mode, expect header is treated like other normal headers */
+    if (expect->base != NULL && h2o_req_should_forward_expect(&conn->req)) {
+        char orig_case[sizeof("expect")] = {0};
+        h2o_memcpy(orig_case, headers[expect_header_index].name, headers[expect_header_index].name_len);
+        h2o_add_header(&conn->req.pool, &conn->req.headers, H2O_TOKEN_EXPECT, orig_case, expect->base, expect->len);
```
```suggestion
        char orig_case_cstr[sizeof("expect")] = {0};
        assert(sizeof(orig_case_cstr) - 1 == headers[expect_header_index].name_len);
        h2o_memcpy(orig_case_cstr, headers[expect_header_index].name, sizeof(orig_case_cstr) - 1);
        h2o_add_header(&conn->req.pool, &conn->req.headers, H2O_TOKEN_EXPECT, orig_case_cstr, expect->base, expect->len);
```
How about this? The intent is
1. to imply that we are converting to cstr, and
2. assert against buffer overrun

kazuho reviewed (commented):
Thank you for the changes. I think this PR is becoming ready (modulo the test for client not waiting for 100).

i110 commented:
@kazuho I fixed an issue that occurs in the following scenario:

1. The client sends a request with `expect: 100-continue`
2. h2o has `proxy.expect: FORWARD`, so the protocol handler doesn’t respond with 100
3. Since client sends no body, `h2o_process_request` doesn’t trigger, and the request is never proxied. This also prevents the upstream from having a chance to respond with 100
4. As a result, it gets stuck (in practice, the client like curl will start sending the body after the wait-100 timeout)

I'll also add some inline comments where I'm not 100% confident.

Also added a test case for `client not waiting for 100` in https://github.com/h2o/h2o/pull/3419/commits/afe07619fb8022d64503057ae6cee3f07668d6e6.

i110 commented on lib/http1.c:
```diff
@@ -662,32 +659,57 @@ static void handle_incoming_request(struct st_h2o_http1_conn_t *conn)
+                    h2o_add_header(&conn->req.pool, &conn->req.headers, H2O_TOKEN_EXPECT, orig_case_cstr, expect_value.base, expect_value.len);
+
+                    h2o_buffer_consume(&conn->sock->input, reqlen);
+                    conn->req.write_req.cb = write_req_connect_first;
```
We should come up with a better name, but nothing came to mind. Any suggestions?

i110 commented on lib/http3/server.c:
```diff
@@ -1548,8 +1548,21 @@ static int handle_input_expect_headers(struct st_h2o_http3_server_stream_t *stre
+            stream->req.proceed_req = proceed_request_streaming;
+            set_state(stream, H2O_HTTP3_SERVER_STREAM_STATE_SEND_HEADERS, 0);
+            quicly_stream_set_receive_window(stream->quic, get_conn(stream)->super.ctx->globalconf->http3.active_stream_window_size);
+            h2o_process_request(&stream->req);
```
I’m not too confident about these lines, so I’d be really happy if you could take a close look!

kazuho replied:
Yeah, `read_blocked` was not set, which could have led to the invocation of `write_req.cb` before the proceed request callback is called.

For h2 and h3, I've merged the logic share with CONNECT.
